keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+* Selected API should support structured logging by providing a way to define a set of key-value pairs for each log entry.
+* There should at least one production ready implementation of the selected API which could produce JSON formatted logs.
+* It should be possible to produce klog compatible logs from the selected API which could be used during the migration to a new API.
+
+Based on this criteria https://github.com/go-logr/logr log interface has been selected as a logging API for Kubernetes.
+
+Apart from meeting the criteria listed above logr has a few nice features:
+* Constant log message together with logger name provide stable identification for log entries which greatly simplifies querying logs.
+* Limiting log levels to just Info and Error makes it much easier to choose the proper log level for given log.
+* Using key-values pairs for defining log fields allows more efficient implementation of the logger.
+
+#### Logger initialization
+
+Logger implementation initialization should be done at startup of component and passed down the stack as an interface. Aside of initialization of logger, there should not be any dependency on specific implementation of logger Interface.
+
+#### Storing Logger and metadata in Context
+
+We would like to propose that logger interface and all metadata collected trough call stack to be stored in Golang Context object. This way creating one canonical way to pass contextual information across the call stack. Some pros of this approach:
+* Standard method in Golang to pass scoped variables across API boundaries and between processes.
+* No need for additional function argument for passing logging API through call stack
+* Context is used similarly by many frameworks including for instance OpenCensus tracing API.
+* Accessing metadata will be independent from Logging and could be reused by other functionalities.
+
+Access to context variables should be done though predefined functions. For that we propose to a new module to kubernetes component base called `k8s.io/component-base/kontext`. This new module will include all utility functions for storing and retrieving logger and metadata needed by it.
+
+Throughout call stack new metadata can be added to context that can be later used by logger. For example reconciliation loop can reference the reconcilided object so that all logs will be referencing it. All metadata access should be done through functions in `k8s.io/component-base/kontext` module. Adding reference to kubernetes object (pod, service, node) should be really straightforward.
+
+#### Example
+
+```go
+import "k8s.io/component-base/kontext"
+
+func UpdateStatus(ctx context.Context, pods []*v1.Pod, newStatus bool) {
+	for _, pod := range pods {
+		podCtx := kontext.WithObject(ctx, pod)
```

Comment:
When, where, and how is this serialized?  These objects are large, so I doubt that anything consuming these want to see them on every message. It seems far more likely that you'd want to write once and reference or something.  Writing a custom handler could handle this, but it seems undesireable to push that change to everyone.

Reply:
We are planning to write to logs only the identification of those contextual objects like name and namespace. Those will be added to all log entries logged using the logger obtained from context object returned from `kontext.WithObject()` function.

Reply:
> We are planning to write to logs only the identification of those contextual objects like name and namespace. Those will be added to all log entries logged using the logger obtained from context object returned from `kontext.WithObject()` function.

Maybe an example will help alleviate concerns, but I suspect many messages today want control of which aspect of objects is dumped and I suspect that most logically have a WithObject (or many) in their context.